chapters/en/chapter10/5.mdx

```diff
@@ -1 +1,362 @@
+
+diagram
+
+2. Decoder: maps the sequence of hidden-states to logits over the vocabulary.
+
+diagram
+
+The encoder is typically of Wav2Vec2 architecture, introduced in Section 4 of this Chapter. We have some flexibility in our choice of decoder. We could either use a simple linear layer that maps the Wav2Vec2 hidden states directly to output logits over our vocabulary. Or, we could pair our Wav2Vec2 encoder with a decoder model, giving a speech encoder-decoder style model, analogous to the NLP encoder-decoder model introduced in Chapter 1.
+
+A simple linear layer will give a smaller, faster overall model, but will be more susceptible to spelling and grammatical errors. Adding a decoder model greatly improves the quality of transcriptions, at the cost of a larger, slower model. This is because the decoder model is pre-trained on a large corpus of text, enabling us to leverage its learned text representations.
+Since we want our system to be robust to spelling and grammar, let’s go ahead and define a speech encoder-decoder style model.
+
+We’ll pair a pre-trained Wav2Vec2 encoder with a pre-trained BART decoder, yielding a Wav2Vev2-2-BART model:
+```python
+from transformers import SpeechEncoderDecoderModel
+
+encoder_id = "facebook/wav2vec2-large-lv60"
+decoder_id = "facebook/bart-large"
+
+model = SpeechEncoderDecoderModel.from_encoder_decoder_pretrained(
+    encoder_id, decoder_id, encoder_add_adapter=True
+)
+
+# set special tokens for generation
+model.config.decoder_start_token_id = model.decoder.config.bos_token_id
+model.config.pad_token_id = model.decoder.config.pad_token_id
+model.config.eos_token_id = model.decoder.config.eos_token_id
+```
+
+The `add_adapter` argument introduces a small convolutional network between the encoder and decoder models. This adapter network helps interface the encoder and decoder by down-sampling the encoder hidden-states to better match the timescale of the decoder. In practice, including this adapter results in superior performance than the encoder and decoder models alone:
+
+diagram
+
+#### Preprocessing the data
+Great! Now that we've defined our model we can start preparing our data for training. For this, we'll need to define two objects: a feature extractor and a tokenizer.
+
+```python
+from transformers import AutoFeatureExtractor, AutoTokenizer, Wav2Vec2Processor
+
+# load feature extractor (for audio inputs) and tokenizer (for text outputs)
+feature_extractor = AutoFeatureExtractor.from_pretrained(encoder_id)
+tokenizer = AutoTokenizer.from_pretrained(decoder_id)
+# combine under one class
+processor = Wav2Vec2Processor(feature_extractor, tokenizer)
+```
+
+We're already pretty familiar with the tokenizer: it converts the text data into a set of token IDs that can be interpreted by the model. This we’ll use to prepare our target labels. But the feature extractor is new! Simply put, the feature extractor prepares our input audio data.
```

Comment:
Would it be more accurate to say input ids instead of token ids?

```suggestion
We're already pretty familiar with the tokenizer: it converts the text data into a set of input IDs that can be interpreted by the model. We’ll use these to prepare our target labels. But the feature extractor is new! Simply put, the feature extractor prepares our input audio data.
```